src/librustc_mir/interpret/memory.rs

```diff
@@ -882,25 +882,16 @@ impl<'a, 'mir, 'tcx, M: Machine<'mir, 'tcx>> Memory<'a, 'mir, 'tcx, M> {
     ) -> EvalResult<'tcx> {
         // The bits have to be saved locally before writing to dest in case src and dest overlap.
```

Comment:
This comment makes me think that we should not do this commit, otherwise we'll run into trouble in the future (and in miri right now). Can you do an `if` for whether there is overlap and if there is, just run the old code?

Reply:
oh right. sorry. I misread the code.

I still think the code isn't doing the right thing. It's only copying once, when it should be copying N-1 times.

You can try this out by creating an array of types with padding, everything starting at the third element will probably not have undef masks for the padding. (you'll need unions to get the bits and then attempt to use them for an array length to actually get a compiler error from that)

Reply:
I'm afraid I'm not quite following. We do call this function with `size * length` so shouldn't it cover all of the repeated copies? Can you provide a sample program that will fail?

Reply:
yes, you are using the length, but that just means that the entire array is copied from `0..N` to `1..=N`, not that the 1st element is copied N times.

I'll make a regression test

Reply:
I'm fairly certain that the following test will succeed to compile on your PR: http://play.rust-lang.org/?gist=1d0183fcfb65164d1ca58ccd9614c33c